File: src/sample/const.ts
type Contains<A> = { a: A, o: any };

type Lens<S, T, A, B> = (f: (a: A) => Contains<B>) => (s: S) => Contains<T>;

function fmap<A, B>(f: (a: A) => B, c: Contains<A>): Contains<B> {
    return { ...c, a: f(c.a) };
}

type Address = { street: string, suburb: string };
type Person = { name: string, address: Address };

const lname : Lens<Person, Person, string, string> = 
    (f: (name: string) => Contains<string>) => 
        (p: Person) => fmap((n => <Person>{ ...p, name: n}), f(p.name));
    
const laddress : Lens<Person, Person, Address, Address> =
    (f: (address: Address) => Contains<Address>) =>
        (p: Person) => fmap((n => <Person>{ ...p, address: n }), f(p.address))

const lstreet : Lens<Address, Address, string, string> =
    (f: (street: string) => Contains<string>) =>
        (a: Address) => fmap((n => <Address>{ ...a, street: n }), f(a.street));

function composeLens<A, B, C>(
    l1: Lens<A, A, B, B>,
    l2: Lens<B, B, C, C>
): Lens<A, A, C, C> {
    return (f: (c: C) => Contains<C>) =>
        (a: A) => l1(b => l2(c => f(c))(b))(a);
}


const p2name = composeLens(laddress, lstreet);

const Identity = <A>(a: A) => <Contains<A>>{ a: a, o: null };
const Const = <A>(a: A) => <Contains<A>>{ a, o: a };
        
const person = { name: 'Hackle', address: { street: 'Sale', suburb: 'CBD' } };
const me = lname(Identity)(person).a;
const justName = lname(Const)(person).o;

console.log(p2name(Identity)(person).a)
console.log(p2name(Const)(person).o)

File: src/sample/const.hs
{-# LANGUAGE RankNTypes #-}

import Data.Functor.Identity
import Data.Functor.Const

type Setter' a b s t = forall f. Functor f => (a -> f b) -> s -> f t

fset :: Setter' a b s t -> (a -> b) -> s -> t
fset set f s = runIdentity $ set (Identity . f) s

_fst :: Setter' a b (a, c) (b, c)
_fst a2fb (a, c) = (\x -> (x, c)) <$> a2fb a

data Address = Address { street :: String, suburb :: String } deriving (Show, Eq)
-- data Person = Person { name :: String, address :: Address } deriving (Show, Eq)
data Person = Person { name :: String } deriving (Show, Eq)

type Lens'' s t a b = forall f. Functor f => (a -> f b) -> s -> f t

lname :: Lens'' Person Person String String
lname fname p = (\n -> p { name = n }) <$> fname (name p)

-- laddress :: Lens'' Person Person Address Address
-- laddress faddress p = (\n -> p { address = n}) <$> faddress (address p)

-- lstreet :: Lens'' Address Address String String
-- lstreet fstreet addr = (\a -> addr { street = a }) <$> fstreet (street addr)

mkIdentity a = (undefined, a)
mkConst a = (a, undefined)
dup a = (a, a)

id1 = runIdentity $ lname Identity $ Person "Hackle"
-- id2 = snd $ lname mkIdentity $ Person "Hackle" (Address "Sale" "CBD")

name1 = getConst $ lname Const $ Person "Hackle"
-- name2 = fst $ lname mkConst $ Person "Hackle" (Address "Sale" "CBD")